transformer_engine/pytorch/module/layernorm_linear.py

```diff
@@ -122,24 +128,27 @@ def forward(
         if ln_bias is not None:
             ln_bias = cast_if_needed(ln_bias, activation_dtype)
 
-        if ub_overlap_ag:
-            raise NotImplementedError
-            tp_world_size = get_distributed_world_size(tp_group)
-            if tp_world_size == 1 or (not is_grad_enabled):
-                ub_overlap_ag = False
-        if ub_overlap_ag:
-            raise NotImplementedError
-            dim_size = list(inputmat.size())
-            dim_size[0] = dim_size[0] * tp_world_size
-            ub_obj_lnout = get_ub(ub_name + "_fprop")
+        tp_world_size = get_distributed_world_size(tp_group)
+        ub_overlap_ag_fprop = (
+            ub_overlap_ag_fprop and is_grad_enabled and not return_layernorm_output
+        )
 
         weight_requires_grad = weight.requires_grad
         backward_needs_input = is_grad_enabled and weight_requires_grad
-        with_input_all_gather = parallel_mode == "column" and sequence_parallel
+        with_input_all_gather = (
+            parallel_mode == "column" and sequence_parallel and not ub_overlap_ag_fprop
+        )
 
         # Configure quantizer for normalization output
-        if fp8 and input_quantizer is None:
-            raise ValueError("Missing quantizer for input tensor")
+        if fp8:
+            if any([ub_overlap_ag_fprop, ub_overlap_rs_fprop]) and isinstance(
+                FP8GlobalStateManager.get_fp8_recipe(), BlockScaling
+            ):
+                raise NotImplementedError("Comm+GEMM overlap does not support MXFP8 block scaling")
+
+            if input_quantizer is None:
+                raise ValueError("Missing quantizer for input tensor")
```

Comment:
- Putting UB logic here makes the comment incorrect
- This won't generalize when we add more quantization schemes. Instead of assuming that all recipes except MXFP8 support UB, we should only assume FP8 delayed scaling supports UB.
```suggestion
        # Check if overlapped communication is supported
        if (
            fp8
            and (ub_overlap_ag_fprop or ub_overlap_rs_fprop)
            and not FP8GlobalStateManager.get_fp8_recipe().delayed()
        ):
            raise NotImplementedError("Comm+GEMM overlap is only supported with FP8 delayed scaling")

        # Configure quantizer for normalization output
        if fp8:
            if input_quantizer is None:
                raise ValueError("Missing quantizer for input tensor")
```